providers/keycloak.go

```diff
@@ -95,5 +128,29 @@ func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.Sess
 		}
 	}
 
+	if len(p.Roles) > 0 {
+		claims := make(map[string]interface{})
+
+		// Decode JWT token without verifying the signature
+		token, err := jwt.ParseSigned(s.AccessToken)
+
+		if err != nil {
+			logger.Printf("failed to parse token %s", err)
+			return "", nil
+		}
+
+		// Parse claims
+		if err := token.UnsafeClaimsWithoutVerification(&claims); err != nil {
+			logger.Printf("failed to parse claims %s", err)
+		}
+
+		var roles = ExtractRolesFromClaims(claims)
+
+		if !isSubarray(roles, p.Roles) {
```

Comment:
So we might need to think about how this aligns with other implementations. In general, in other group/role AuthZ logic in other providers, we've made this a if any roles match the list, access is allowed (instead of all must be present).

Reply:
Oh, are roles being used in any other provider? The only thing I've seen is groups, and for that it makes sense that access would be granted for a user belonging to any specified group. This was based on the Louketo Proxy implementation.

> By default, the roles defined on a resource perform a logical AND so all roles specified must be present in the claims, this behavior can be altered by the require-any-role option, however, so as long as one role is present the permission is granted.

I am open for suggestions though, would you like me to change it so that it grants access if any of the provided roles are present?